Implement capabilities endpoint

lindholc opened the pull request:
I've implemented the [capabilities endpoint](https://github.com/hapi-server/data-specification/blob/master/hapi-2.0.0/HAPI-data-access-spec-2.0.0.md#capabilities) with the assumption that we will support all of the standard HAPI outputs.

I imagine in the future we might revisit this so the list of outputs is dynamically generated and includes custom output formats. (Though it isn't clear to me how the custom formats are expected to behave.)

I've also left these two commits separate for the purpose of this review:
- [6a014c0](https://github.com/lasp/latis-hapi/pull/21/commits/6a014c0f107f28e25d47268224779fe8da639d3a) : tests implemented with [uTest](https://github.com/lihaoyi/utest)
- [38fec45](https://github.com/lasp/latis-hapi/pull/21/commits/38fec45654ae505d1a7f988048eff5cf31fb6f6f) : tests implemented with [ScalaTest](http://www.scalatest.org/)

I'm curious what you think about the two libraries. uTest is meant to be more minimal, which appeals to me. However, ScalaTest is pretty widely used.

Resolves #4.

lindholc commented on src/main/scala/latis/server/CapabilitiesService.scala:
```diff
@@ -10,6 +13,30 @@ class CapabilitiesService[F[_]: Effect] extends Http4sDsl[F] {
+final case class Capabilities(
+  version: String,
+  status: Status,
+  formats: List[String]
```
Note for the future:

I can imagine defining something like

```scala
sealed abstract trait OutputFormat
final case object CSV extends OutputFormat
final case object JSON extends OutputFormat
final case object BIN extends OutputFormat
final case class Custom(name: String) extends OutputFormat
```

This is probably better than accepting arbitrary strings, and we could enforce the requirement that custom output formats start with `x_` in the response.

lindholc commented on src/main/scala/latis/server/CapabilitiesService.scala:
```diff
@@ -10,6 +13,30 @@ class CapabilitiesService[F[_]: Effect] extends Http4sDsl[F] {
       case GET -> Root / "capabilities" =>
-        Ok("Hello from HAPI!")
+        val capabilities =
+          Capabilities("2.0", Status.`1200`, List("binary", "csv", "json"))
```
I think I'll factor out the version of the HAPI spec. I told myself I'd do that when I used the version in more than one place, but I already am!

lindholc commented on build.sbt:
```diff
@@ -11,10 +11,13 @@ lazy val `latis-hapi` = (project in file("."))
+      "com.lihaoyi"   %% "scalatags"           % "0.6.7",
+      "org.scalatest" %% "scalatest"           % "3.0.5" % Test
+    ),
+    Test / logBuffered := false
```
[ScalaTest recommends this setting](http://www.scalatest.org/user_guide/using_scalatest_with_sbt) because they do their own buffering.

lindholc commented on src/test/scala/latis/server/CapabilitiesServiceSpec.scala:
```diff
@@ -0,0 +1,37 @@
+
+class CapabilitiesServiceSpec extends FlatSpec {
+
+  "The capabilities service" should "advertise all output options" in {
```
ScalaTest leans towards BDD-style tests, which is different from the JUnit-style tests we've been writing.

I like how this forces me to write out what ought to be happening, which gives more structure and direction to the test because I need to be able to describe it succinctly. However, I can also see this being difficult.

ScalaTest lets you combine multiple styles of testing, so maybe this isn't an issue. There could be specification-type tests and JUnit-type tests for things like reproducing a particular bug.

dlindhol replied:
I use snake_case for test names so I can be more verbose, a recommendation I picked up somewhere. I can see this becoming a bit onerous though it makes things more clear. I can always cut and paste.

The one think I want back is the ability to run a single test (not an entire class) in eclipse. I lost that with JUnit after a recent upgrade.

lindholc commented on src/test/scala/latis/server/CapabilitiesServiceSpec.scala:
```diff
@@ -0,0 +1,37 @@
+    val capabilities = new CapabilitiesService[IO]()
+
+    // The stream ought to contain only the body.
+    val (body :: Nil) =
```
This is normally a bad thing to do because this will fail if the list is empty or has more than one thing in it. I'm surprised there's not a warning about it.

But, because this is a test, I think it's fine.

dlindhol replied:
I wouldn't worry about totality of code in tests. If the test fails due to an assumed result (e.g. bad pattern match) then it still fails. We may need to relax some compiler warnings on test code.

lindholc commented:
I've fixed up my history and chose the ScalaTest commit for the tests.

I've also added a single place for the version of the HAPI spec we're implementing.
